### ISSUE TITLE
No securityContext settings for initHelper

### ISSUE THREAD

**Issue by jbnjohnathan** [Opster/opensearch-k8s-operator#598]
I am trying to deploy OpenSearch using this operator in a OpenShift cluster.
The cluster has constraints on securityContext which means that all pods needs to have specific securityContext settings.
I configured this for the opensearch and opensearch dashboard within the CR and then deployed the CR.
No pods came up, so I checked the operator logs

```
1.6939888173352401e+09  ERROR   Reconciler error        {"controller": "opensearchcluster", "controllerGroup": "opensearch.opster.io", "controllerKind": "OpenSearchCluster", "OpenSearchCluster": {"name":"opensearch","namespace":"XXX"}, "namespace": "XXX", "name": "opensearch", "reconcileID": "XXX", "error": "failed to create resource: creating resource failed: pods \"opensearch-bootstrap-0\" is forbidden: unable to validate against any security context constraint: [provider \"anyuid\": Forbidden: not usable by user or serviceaccount, spec.initContainers[0].securityContext.runAsUser: Invalid value: 0: must be in the ranges: [1000800000, 1000809999], spec.initContainers[1].securityContext.privileged: Invalid value: true: Privileged containers are not allowed, provider \"restricted\": Forbidden: not usable by user or serviceaccount, provider \"nonroot-v2\": Forbidden: not usable by user or serviceaccount, provider \"nonroot\": Forbidden: not usable by user or serviceaccount, provider \"hostmount-anyuid\": Forbidden: not usable by user or serviceaccount, provider \"machine-api-termination-handler\": Forbidden: not usable by user or serviceaccount, provider \"hostnetwork-v2\": Forbidden: not usable by user or serviceaccount, provider \"hostnetwork\": Forbidden: not usable by user or serviceaccount, provider \"hostaccess\": Forbidden: not usable by user or serviceaccount, provider \"node-exporter\": Forbidden: not usable by user or serviceaccount, provider \"privileged\": Forbidden: not usable by user or serviceaccount]", "errorVerbose": "pods \"opensearch-bootstrap-0\" is forbidden: unable to validate against any security context constraint: [provider \"anyuid\": Forbidden: not usable by user or serviceaccount, spec.initContainers[0].securityContext.runAsUser: Invalid value: 0: must be in the ranges: [1000800000, 1000809999], spec.initContainers[1].securityContext.privileged: Invalid value: true: Privileged containers are not allowed, provider \"restricted\": Forbidden: not usable by user or serviceaccount, provider \"nonroot-v2\": Forbidden: not usable by user or serviceaccount, provider \"nonroot\": Forbidden: not usable by user or serviceaccount, provider \"hostmount-anyuid\": Forbidden: not usable by user or serviceaccount, provider \"mac
```

I checked the CRD and could not find any securityContext settings for the `initHelper` (bootstrap pod). I only found securityContext settings for the cluster and dashboard.

**Comment by swoehrl-mw** [comment]
Hi @jbnjohnathan. You are correct that it is currently not possible to configure a securitycontext for the init container. But this could be problematic as the init container needs to run as root to do a chown on the data.
For situations like yours we have an option to disable the init container: https://github.com/Opster/opensearch-k8s-operator/blob/main/docs/userguide/main.md#disabling-the-init-helper
Does that work for you?

**Comment by jbnjohnathan** [comment]
Thanks for your swift reply! 
It could work, but I'm not sure if I currently understand exactly what purpose the init container running the chown have, and what the implications of disabling it will be.
If you can let me know I could see if the lack of it poses an issue.
Thanks!

**Comment by swoehrl-mw** [comment]
I'm not an expert on this part of opensearch but afaik the chown is there in case permissions on the volume are wrong (for PVCs they should be automatically set correctly by Kubernetes itself but for other volume types they might not). So you should be fine disabling the init container in most cases.

**Comment by jbnjohnathan** [comment]
@swoehrl-mw I disabled the initHelper container and redeployed.
Now I get the same error on the `opensearch-bootstrap` pod. 
For this there is a section called "additionalConfig" that might be able to be used to set securityContext on the bootstrap pod, but I am not sure.

**Comment by swoehrl-mw** [comment]
@jbnjohnathan  `additionalConfig` is only for config that would go into `opensearch.yaml`. For the bootstrap pod the securityContext under `spec.general.securityContext` and `spec.general.podSecurityContext` are used. Have you configured those?

Please also make sure you are using the new 2.4.0 version of the operator. There was an enhancement that added securitycontext to the securityconfig-update job.

**Comment by jbnjohnathan** [comment]
@swoehrl-mw I tried upgrading to 2.4.0 but the problem persists.
I checked through the code and cannot see that the spec.general.podSecurityContext would apply to the bootstrap pod.

**Comment by swoehrl-mw** [comment]
@jbnjohnathan 

It is definitly in the code:
https://github.com/Opster/opensearch-k8s-operator/blob/5666be7749e43d404ea4fa301361e37a51eedb0e/opensearch-operator/pkg/builders/cluster.go#L696
https://github.com/Opster/opensearch-k8s-operator/blob/5666be7749e43d404ea4fa301361e37a51eedb0e/opensearch-operator/pkg/builders/cluster.go#L806

Can you please provide a (minimal) OpensearchCluster definition to reproduce?

**Comment by jbnjohnathan** [comment]
Ah interesting. It does indeed seem like it would work.
Let me try again and give some more details to help out.

First I deploy the operator with the image `opensearch-operator:2.4.0`
In the operator I have set securityContext so it deploys without issues.

Next lets deploy a manifest of kind `OpenSearchCluster`
Since the helm one is riddled with variables, I have now created a "clean" one with static information. Or rather, I used `helm template` to generate the same manifest I have tested with before

```yaml
---
# Source: opensearch/charts/opensearch-cluster/templates/opensearch-cluster-cr.yaml
apiVersion: opensearch.opster.io/v1
kind: OpenSearchCluster
metadata:
  name: opensearch
  namespace: mynamespace
spec:
  initHelper:
    version: 1.27.2-buildx

    image: private-repo/mynamespace/busybox
  general:
    version: 2.8.0
    image: "private-repo/mynamespace/opensearch"
    httpPort: 9200
    vendor: opensearch
    serviceName: opensearch-http
    setVMMaxMapCount: true
    drainDataNodes: true
    securityContext:
      privileged: false
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      readOnlyRootFilesystem: true
      runAsNonRoot: true
      seccompProfile:
        type: RuntimeDefault
  dashboards:
    image: "private-repo/mynamespace/opensearch-dashboards"
    version: 2.8.0
    enable: true
    replicas: 1
    resources:
      requests:
        memory: 1Gi
        cpu: 500m
      limits:
        memory: 1Gi
        cpu: 500m
    securityContext:
      privileged: false
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      readOnlyRootFilesystem: true
      runAsNonRoot: true
      seccompProfile:
        type: RuntimeDefault
  nodePools:
    - component: masters
      replicas: 1
      diskSize: "30Gi"
      resources:
        requests:
          memory: 2Gi
          cpu: 500m
        limits:
          memory: 2Gi
          cpu: 500m
      roles:

      - master
      - data
  security:
    tls:
      transport:
        generate: true
      http:
        generate: true
```

I deployed this manifest, and then checked the operator pod for errors
There was no opensearch deployed, and I got the following error from the operator
```
{"level":"error","ts":"2023-09-14T11:25:18.527Z","msg":"Reconciler error","controller":"opensearchcluster","controllerGroup":"opensearch.opster.io","controllerKind":"OpenSearchCluster","OpenSearchCluster":{"name":"opensearch","namespace":"mynamespace"},"namespace":"mynamespace","name":"opensearch","reconcileID":"6ce5cfb5-f315-45ad-a93f-d9b966089745","error":"failed to create resource: creating resource failed: pods \"opensearch-bootstrap-0\" is forbidden: unable to validate against any security context constraint: [provider \"anyuid\": Forbidden: not usable by user or serviceaccount, spec.initContainers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed, provider \"restricted\": Forbidden: not usable by user or serviceaccount, provider \"nonroot-v2\": Forbidden: not usable by user or serviceaccount, provider \"nonroot\": Forbidden: not usable by user or serviceaccount, provider \"hostmount-anyuid\": Forbidden: not usable by user or serviceaccount, provider \"machine-api-termination-handler\": Forbidden: not usable by user or serviceaccount, provider \"hostnetwork-v2\": Forbidden: not usable by user or serviceaccount, provider \"hostnetwork\": Forbidden: not usable by user or serviceaccount, provider \"hostaccess\": Forbidden: not usable by user or serviceaccount, provider \"node-exporter\": Forbidden: not usable by user or serviceaccount, provider \"privileged\": Forbidden: not usable by user or serviceaccount]","errorVerbose":"pods \"opensearch-bootstrap-0\" is forbidden: unable to validate against any security context constraint: [provider \"anyuid\": Forbidden: not usable by user or serviceaccount, spec.initContainers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed, provider \"restricted\": Forbidden: not usable by user or serviceaccount, provider \"nonroot-v2\": Forbidden: not usable by user or serviceaccount, provider \"nonroot\": Forbidden: not usable by user or serviceaccount, provider \"hostmount-anyuid\": Forbidden: not usable by user or serviceaccount, provider \"machine-api-termination-handler\": Forbidden: not usable by user or serviceaccount, provider \"hostnetwork-v2\": Forbidden: not usable by user or serviceaccount, provider \"hostnetwork\": Forbidden: not usable by user or serviceaccount, provider \"hostaccess\": Forbidden: not usable by user or serviceaccount, provider \"node-exporter\": Forbidden: not usable by user or serviceaccount, provider \"privileged\": Forbidden: not usable by user or serviceaccount]\ncreating resource failed\ngithub.com/cisco-open/operator-tools/pkg/reconciler.(*GenericResourceReconciler).CreateIfNotExist\n\t/go/pkg/mod/github.com/cisco-open/operator-tools@v0.30.0/pkg/reconciler/resource.go:594\ngithub.com/cisco-open/operator-tools/pkg/reconciler.(*GenericResourceReconciler).ReconcileResource\n\t/go/pkg/mod/github.com/cisco-open/operator-tools@v0.30.0/pkg/reconciler/resource.go:382\nopensearch.opster.io/pkg/reconcilers.(*ClusterReconciler).Reconcile\n\t/workspace/pkg/reconcilers/cluster.go:113\nopensearch.opster.io/controllers.(*OpenSearchClusterReconciler).reconcilePhaseRunning\n\t/workspace/controllers/opensearchController.go:321\nopensearch.opster.io/controllers.(*OpenSearchClusterReconciler).Reconcile\n\t/workspace/controllers/opensearchController.go:142\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1594\nfailed to create resource","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226"}
```

Some of the error messages from the above wall of text
```
Failed to create resource: creating resource failed: pods \"opensearch-bootstrap-0\" is forbidden: unable to validate against any security context constraint: [provider \"anyuid\": Forbidden: not usable by user or serviceaccount, spec.initContainers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed
```

privilegied is default `false` because it gives the pod a lot of scary access to the kubernetes host.
I noticed that even if I didn't set `spec.general.privileged` it defaults to `true`
But even when I set it to `false` it is set to `true`

I checked the code and found this line. I don't really know goland but it looks like its always set to `true`

https://github.com/Opster/opensearch-k8s-operator/blob/5666be7749e43d404ea4fa301361e37a51eedb0e/opensearch-operator/pkg/builders/cluster.go#L482

**Comment by swoehrl-mw** [comment]
> I think I see why its privilegied. It is because of setVMMaxMapCount was set to true. This triggered the init-sysctl init container which always set privilegied to true which is not allowed in our cluster.

Correct. That init container can only run with priviledged access (because it uses sysctl), but it can easily be disabled  (via `setVMMaxMapCount: false`) if your hosts are already properly configured. It's more of a convienence helper to not have to deal with this setting separately.

Does it work if you disable `setVMMaxMapCount`?

**Comment by jbnjohnathan** [comment]
Yes setting that to false solves the issue. 
Maybe this should be better reflected in the documentation. 
That spec.general.securityContext is inherited down to the bootstrap pod as well as the bootstrap init pod apart from the privilege value.

**Comment by swoehrl-mw** [comment]
@jbnjohnathan I've added a paragraph to the userguide to make the init container behaviour clearer in regards to the security context (https://github.com/Opster/opensearch-k8s-operator/commit/7bdbfc08fe92193ad792508f27ea5eff1eefdfc4).

Closing this issue as completed.
